### PR TITLE
Severely extended the sorin theme

### PR DESCRIPTION
**Pull request by lhecker**
## Description

As I mentioned in #5388 I was relatively unhappy with the sorin theme due it's numerous faults in its git handling.
It also didn't really match the original anywhere exactly – If the symbols where right, the colors where off; If the colors where right, the data was invalid. And so on...

Thus I pretty much rewrote the whole thing from scratch. The theme should now mostly match the original. :slightly_smiling_face: 
(Even though nothing from the original is left I didn't set myself as the author, because I wasn't sure if I may or if that'd be beneficial.)

As mentioned before I'm relatively new to fish and there might be errors in the code.
But I'm eager to hear what you think about this PR. :slightly_smiling_face: 
And if you can find the time I'd be great if you could try out the theme and tell me whether you happen to encounter any symbol errors for the `fish_right_prompt`.

## Features

...compared to the previous version:
* Correct, bright colors and bold fonts wherever needed
* Fixed symbols for all remaining git states (e.g. arrows signaling whether you're behind/ahead of the origin, or whether anything is stashed)
* Fixed extraction of the branch names and commits refs

...compared to all other themes (as far as I can see):
* The parsing of `git status --porcelain` now works correctly in all cases, thanks to prezto's helpful documentation. `acidhub` and `__terlar_git_prompt` do it incorrectly, [because the git status identification is quite a bit more nuanced than you might think](http://i.imgur.com/2YLu1.png). (I was surprised too!)

## TODOs:
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

P.S.: Today I found #540. It's very disheartening how that issue was handled, especially since quite a few people had quite valid, subjective reasons to be using heredocs. I had the issue myself today as I tried printing a large string consisting of single- and doublequotes, while working on this PR, but there's nothing for fish allowing you to simply paste the string. No, instead you need to replace all quotes every time. Something like C++'s raw string literals would've been nice. 😕

### CODE REVIEW

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 1]
```diff
@@ -1,61 +1,219 @@
...
+    # Return if zprezto's git-info is disabled.
+    set -l git_info_enabled (command git config --bool prompt.showinfo)
+    if test "$git_info_enabled" = 'false'
+        set_color normal
+        return
+    end
+
+    set -l commit ''
+    if set -l action (_git_action "$git_dir")
+        set commit (command git rev-parse --short=7 HEAD 2> /dev/null)
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: See 8e88b29eeb9aa7c42a00a4188c6c26a17d8ad9f5. Git's shortening can be quite slow, because it tries to be unambiguous. For a prompt, that's overkill.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 2]
```diff
@@ -1,61 +1,219 @@
...
+
+    # Get either the branch name or a branch descriptor.
+    set -l branch_detached 0
+    if not set -l branch (command git symbolic-ref --short HEAD ^&-)
+        set branch_detached 1
+        set branch (command git describe --contains --all HEAD ^&-)
+    end
+
+    # Get the commit difference counts between local and remote.
+    set -l IFS \t
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: This is fish 3.0. Please use `read -d \t` instead of messing with $IFS.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 3]
```diff
@@ -1,61 +1,219 @@
...
-                printf (set_color magenta)⬇' '
-            end
+    set -l cmd_status $status
+    if test $cmd_status -ne 0
+        echo -n (set_color red)"✘ $cmd_status "
+    end
+
+    # Get the git directory for later use.
+    # Return if not inside a Git repository work tree.
+    set -l git_dir (command git rev-parse --git-dir ^&-)
```
[the remark this answers is not in the document]
Comment (faho) [reply share/tools/web_config/sample_prompts/sorin.fish]: Also the `^` redirection has been deprecated. Use `2>/dev/null`.

You usually also don't want to close, some programs handle that badly.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 4]
```diff
@@ -1,61 +1,219 @@
...
+        set status_behind 0
+    end
+
+    # Get the stash status.
+    # (git stash list) is very slow. => Avoid using it.
+    set -l status_stashed 0
+    if test -f "$git_dir/refs/stash"
+        set status_stashed 1
+    else if test -r "$git_dir/commondir"
+        set -l IFS \n
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: If you don't do the IFS setting before, this can be removed.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 5]
```diff
@@ -1,61 +1,219 @@
...
-                    printf (set_color blue)✱' '
-                case "*R*"
-                    printf (set_color brmagenta)➜' '
-                case "*U*"
-                    printf (set_color bryellow)═' '
-                case "??"
-                    printf (set_color brwhite)◼' '
+    # Count added, deleted, modified, renamed, unmerged, and untracked.
+    # T (type change) is undocumented, see http://git.io/FnpMGw.
+    # For a table of scenarii, see http://i.imgur.com/2YLu1.png.
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: That's gonna bitrot.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 6]
```diff
@@ -1,61 +1,219 @@
...
+    set -l status_added 0
+    set -l status_deleted 0
+    set -l status_modified 0
+    set -l status_renamed 0
+    set -l status_unmerged 0
+    set -l status_untracked 0
+    for line in (command git status --porcelain | cut -b-2)
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: `string sub -l 2` instead of cut.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 7]
```diff
@@ -1,61 +1,219 @@
...
+        echo -n ' '(set_color yellow)'═'
+    end
+    if test $status_untracked -ne 0
+        echo -n ' '(set_color white)'◼'
+    end
+
+    set_color normal
+end
+
+function _git_action --argument-names git_dir
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: This could probably be extracted as a helper function.

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 8]
```diff
@@ -1,61 +1,219 @@
...
+    for action_dir in "$git_dir/rebase-apply" "$git_dir/rebase"
+        if test -d "$action_dir"
+            if test -f "$action_dir/rebasing"
+                echo -n 'rebase'
```
Comment (faho) [inline share/tools/web_config/sample_prompts/sorin.fish]: The quotes here around literals are unnecessary. Not that they hurt, of course.

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 4 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: What if the user sets the IFS variable? Wouldn't that affect the theme as well?

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 5 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: Do you mean the situation if imgur is gone in the future? The link itself stems from prezto's source code.
What about if I create a ASCII-style table inside the theme's source code? Would that work for you?
The resulting comment would be extremely wide though of course.

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 8 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: Would you like me to remove them?

File: share/tools/web_config/sample_prompts/sorin.fish [hunk 9]
```diff
@@ -1,61 +1,219 @@
...
+
+    # Get the git directory for later use.
+    # Return if not inside a Git repository work tree.
+    set -l git_dir (command git rev-parse --git-dir ^&-)
+    if test $status -ne 0
+        set_color normal
+        return
+    end
+
+    # Return if zprezto's git-info is disabled.
```
[the remark this answers is not in the document]
Comment (faho) [reply share/tools/web_config/sample_prompts/sorin.fish]: My concern here is that this is one of those options nobody will find.

Though injecting it into git config is kind of clever, and allows you to e.g. not run the full informative status on large repos.

On the other hand, this option here disables _everything_.  Getting e.g. the SHA for HEAD and the current action is quite quick even on large repositories (as long as you don't need to disambiguate), so I really don't see a need for this in particular. The `git status` call will dwarf everything else.

We can use other projects' settings or add our own, but:

- This one seems less than optimal
- This is just in one prompt and not even in our main functions/__fish_git_prompt
- It should be documented

So, I'd like for this to be removed _now_, and further changes to be discussed and figured out.

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 7 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: I extracted it into `__fish_git_action` now. Do you prefer this?
Or would you maybe like to rename the function?

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 4 above)
Comment (faho) [reply share/tools/web_config/sample_prompts/sorin.fish]: If the user sets IFS globally, then yes, it will break.

But then everything will break and the user gets to keep the pieces.

This isn't POSIX sh, our IFS isn't really the same thing (even though it looks like it). Since we don't do word-splitting, we really only use $IFS in two places:

- `read`, which now has a "-d"/"--delimiter=" option that takes precedence.
- Command substitutions, where we don't use the _actual value_, but only if the variable is set (in which case we split on newlines) or not (in which case we don't split). This can be better done via `string split0`, which will cause the command substitution to split on NULs (which can't be passed as arguments anyway).

We want to move away from $IFS - _because_ it can be set globally - which is why we've introduced these alternatives in fish 3.

Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: > But then everything will break and the user gets to keep the pieces.
I personally can't agree with that. IMO, since a theme is just that, it should be robust against all (or almost all) commands a user enters without breaking.

As such... Would you mind if I keep this as it is?
It already uses the `-d \n` flag you suggested and it shouldn't make a difference to fish if the flag is there or not, right? :slightly_smiling_face: 
Basically I'd just like to be explicit here: I want to read an entire line (potentially the whole file) and not just a whitespace-separated word.

If you have another idea how to write this code more elegantly I'm all for it though. :slightly_smiling_face:

**Comment by floam** [comment]
If you've rewritten this, my take is that it's OK to set yourself as the author.

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 7 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: Hmm... I was just thinking about this and...
I believe it might be better to keep it as a private function, because as soon as it isn't anymore, you need to keep backwards-compatibility in mind, right?
And I'm not sure if we should push for that already, since this function is kinda... specialized, unlike `fish_print_hg_root`.

What's your opinion on this?

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 5 above)
Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: Done. I was able to fit the whole thing within 80 characters. :+1:

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 7 above)
Comment (faho) [reply share/tools/web_config/sample_prompts/sorin.fish]: >this function is kinda... specialized, unlike fish_print_hg_root.

Is it? I mean it's mostly translating git's "API" into strings. You'd have to keep those constant for any users, but that's not that huge a burden.

Absolute worst case? You translate them again in the prompt. E.g. we could replace https://github.com/fish-shell/fish-shell/blob/master/share/functions/__fish_git_prompt.fish#L562-L592 with something like

```fish
switch (fish_git_action)
    case rebase
        echo "|REBASE"
    case merge
        echo "|MERGING"
    # and so on
end
```

And you'd still isolate a bunch of knowledge about git in one function.

And that prompt is a nice example, because `git blame -w` tells us that, ignoring whitespace-only changes, those strings haven't changed since 2013.

**Comment by lhecker** [comment]
I believe we should extract the git status handling into a seperate function as well.
This would allow other themes to profit from the more robust procelain status parsing.
What do you think? @faho

**Comment by faho** [comment]
>I believe we should extract the git status handling into a seperate function as well.
This would allow other themes to profit from the more robust procelain status parsing.

Sure. But that would require a robust "protocol", because you need to send a count of files etc.

It's a great idea, but it should be a separate PR.

**Comment by lhecker** [comment]
> But that would require a robust "protocol", because you need to send a count of files etc.

That's actually exactly why I meant that it should be a private `__fish_git_action` function!

But basically I had in mind to combine the logic of `__terlar_git_prompt` with this theme's and add a function that returns a list of human-readable names (like "added", "deleted", "untracked", ...) and switch-case over it in the themes. A set/map type would be kinda useful here as mentioned elsewhere. 😅
I don't disagree with your assessment though.

**Comment by lhecker** [comment]
Please let me know if I you'd like me to change anything else in this PR. :slightly_smiling_face:

**Comment by faho** [comment]
That's some quality work, thanks!

And it's merged, so it will be in 3.1.

File: share/tools/web_config/sample_prompts/sorin.fish (hunk 7 above)
Comment (comfortablynick) [reply share/tools/web_config/sample_prompts/sorin.fish]: > I named it `fish_print_git_action` to be a bit analogue to your hq command.
> 

I just tried it from today's build and it looks like `fish_right_prompt` is still calling `__fish_git_action`. Once I changed it to `fish_print_git_action`, it worked fine.

Comment (faho) [reply share/tools/web_config/sample_prompts/sorin.fish]: Fixed in master, sorry!

Comment (lhecker) [reply share/tools/web_config/sample_prompts/sorin.fish]: Thanks for taking care of this @faho. 🙂
